core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java

```diff
@@ -336,12 +340,31 @@ public final void close() {
 
     @Override
     public String toString() {
+        return toStringHelper().toString();
+    }
+
+    protected ToStringHelper toStringHelper() {
         return MoreObjects.toStringHelper(this)
+                          .omitNullValues()
                           .add("selectionStrategy", selectionStrategy.getClass())
                           .add("allowsEmptyEndpoints", allowEmptyEndpoints)
                           .add("endpoints", truncate(endpoints, 10))
                           .add("numEndpoints", endpoints.size())
-                          .add("initialized", initialEndpointsFuture.isDone())
-                          .toString();
+                          .add("initialized", initialEndpointsFuture.isDone());
+    }
+
+    private class InitialEndpointsFuture extends EventLoopCheckingFuture<List<Endpoint>> {
+
+        @Override
+        public List<Endpoint> get(long timeout, TimeUnit unit)
+                throws InterruptedException, ExecutionException, TimeoutException {
+            try {
+                return super.get(timeout, unit);
+            } catch (TimeoutException e) {
+                throw new TimeoutException(
+                        InitialEndpointsFuture.class.getSimpleName() + " is timed out after " +
+                        unit.toMillis(timeout) + " milliseconds. endpoint group: " + DynamicEndpointGroup.this);
```

Comment:
How about preserving the original `TimeoutException`'s stack traces?
```suggestion
                final TimeoutException timeoutException = new TimeoutException(
                        InitialEndpointsFuture.class.getSimpleName() + " is timed out after " +
                        unit.toMillis(timeout) + " milliseconds. endpoint group: " + DynamicEndpointGroup.this);
                timeoutException.initCause(e);
                throw timeoutException;
```